src/compiler/binder.ts

```diff
@@ -1751,4 +1791,440 @@ namespace ts {
+                // A ClassDeclarations or ClassExpression is ES6 syntax.
+                excludeFlags = TransformFlags.ClassExcludes;
+                transformFlags = TransformFlags.AssertES6;
+
+                // A class with a parameter property assignment, property initializer, or decorator is
+                // TypeScript syntax.
+                if (subtreeFlags & TransformFlags.ContainsParameterPropertyAssignments
+                    || subtreeFlags & TransformFlags.ContainsPropertyInitializer
+                    || subtreeFlags & TransformFlags.ContainsDecorators) {
+                    transformFlags |= TransformFlags.AssertTypeScript;
+                }
+
+                break;
+
+            case SyntaxKind.HeritageClause:
+                if ((<HeritageClause>node).token === SyntaxKind.ExtendsKeyword) {
+                    // An `extends` HeritageClause is ES6 syntax.
+                    transformFlags |= TransformFlags.AssertES6;
+                }
+                else {
+                    // An `implements` HeritageClause is TypeScript syntax.
+                    Debug.assert((<HeritageClause>node).token === SyntaxKind.ImplementsKeyword);
+                    transformFlags |= TransformFlags.AssertTypeScript;
+                }
+
+                break;
+
+            case SyntaxKind.ExpressionWithTypeArguments:
+                // An ExpressionWithTypeArguments is ES6 syntax, as it is used in the
+                // extends clause of a class.
+                transformFlags |= TransformFlags.AssertES6;
+
+                // If an ExpressionWithTypeArguments contains type arguments, then it
+                // is TypeScript syntax.
+                if ((<ExpressionWithTypeArguments>node).typeArguments) {
+                    transformFlags |= TransformFlags.AssertTypeScript;
+                }
+
+                break;
+
+            case SyntaxKind.Constructor:
+                // A Constructor is ES6 syntax.
+                excludeFlags = TransformFlags.ConstructorExcludes;
+                transformFlags |= TransformFlags.AssertES6;
+
+                // An overload constructor is TypeScript syntax.
+                if (!(<ConstructorDeclaration>node).body) {
+                    transformFlags |= TransformFlags.AssertTypeScript;
+                }
+
+                break;
+
+            case SyntaxKind.PropertyDeclaration:
+                // A PropertyDeclaration is TypeScript syntax.
+                transformFlags |= TransformFlags.AssertTypeScript;
+
+                // If the PropertyDeclaration has an initializer, we need to inform its ancestor
+                // so that it handle the transformation.
+                if ((<PropertyDeclaration>node).initializer) {
+                    transformFlags |= TransformFlags.ContainsPropertyInitializer;
+                }
+
+                break;
+
+            case SyntaxKind.MethodDeclaration:
+                // A MethodDeclaration is ES6 syntax.

```

Comment:
you don't need to capture whether the method is static or instance? Wouldn't that affect the emitting?